### PR TITLE
Feature/language switcher

### LINKED ISSUE #108: Change language based on a setting
### Proposed Solution

Add language setting that overrides the browser language; maybe via a cookie?

### PR DESCRIPTION
**Pull request by noworneverev**
### Short description
User now is able to change languages based on their selection of language.

### Proposed changes
Add a switch button in sidebar to let user switch between German and English.
- 
- 

### Side effects

- 
- 

### Resolved issues

Fixes: #108

### CODE REVIEW

File: ycms/cms/templates/_base.html [hunk 1]
```diff
@@ -111,25 +111,44 @@
...
+                 class="absolute z-10 invisible inline-block px-3 py-2 text-sm font-medium text-white bg-gray-800 rounded-lg shadow-sm opacity-0 tooltip">
+                {% if is_dark_theme %}
+                    {% translate "Turn on the light" %}
+                {% else %}
+                    {% translate "Turn off the light" %}
+                {% endif %}
+                <div class="tooltip-arrow" data-popper-arrow></div>
+            </div>
+            <div class="flex items-center ml-4 mb-2">
+                <form class="mr-2" action="{% url 'switch-language' 'en' %}" method="post">
```
Comment (charludo) [inline ycms/cms/templates/_base.html]:
```suggestion
        <div class="flex justify-center items-center mb-4 gap-4">
            <a href="{% url 'change-theme' %}"
               data-tooltip-target="tooltip-right"
               data-tooltip-placement="right"
               class="inline-block text-gray-500 dark:text-gray-400 hover:bg-gray-100 dark:hover:bg-gray-600 focus:outline-none focus:ring-4 focus:ring-gray-200 dark:focus:ring-gray-800 rounded-lg text-sm p-2.5">
                {% if is_dark_theme %}
                    <i icon-name="sun"></i>
                {% else %}
                    <i icon-name="moon"></i>
                {% endif %}
            </a>
            <div id="tooltip-right"
                 role="tooltip"
                 class="absolute z-10 invisible inline-block px-3 py-2 text-sm font-medium text-white bg-gray-800 rounded-lg shadow-sm opacity-0 tooltip">
                {% if is_dark_theme %}
                    {% translate "Turn on the light" %}
                {% else %}
                    {% translate "Turn off the light" %}
                {% endif %}
                <div class="tooltip-arrow" data-popper-arrow></div>
            </div>
            <div class="flex items-center">
                <form class="mr-2" action="{% url 'switch-language' 'en' %}" method="post">
```
It's currently a bit unaligned, this fixes it for me.

**Review by charludo — CHANGES_REQUESTED** [review]
Cool, this works perfectly! :tada:

My only issue is that when you visit the page from a new device for the first time, no language option is selected. This might confuse users a little, esp. if we're going to host a public instance for the presentation, where they will be able to use their own devices.

Maybe something like this could fix this, by just setting the cookie based on the browser settings if it is not yet set?

```typescript
const getUserBrowserLanguage = (): string => {
    const userLanguage = navigator.language.split('-')[0];
    return ['de', 'en'].includes(userLanguage) ? userLanguage : 'en';
};

const checkLanguageCookie = (): boolean => !!document.cookie.split('; ').find((cookie) => cookie.startsWith('django_language='));

const setLanguageCookie = (language: string): void => {
    document.cookie = `django_language=${language}; path=/;`;
};

const addClassesToElement = (elementId: string, classes: string): void => {
    const element = document.getElementById(elementId);
    if (element) {
        element.classList.add(...classes.split(' '));
    }
};

const handleLanguageSwitch = (): void => {
    if (!checkLanguageCookie()) {
        const userLanguage = getUserBrowserLanguage();
        setLanguageCookie(userLanguage);

        const elementId = `lang-switch-${userLanguage}`;
        const classesToAdd = 'underline pointer-events-none text-gray-300 dark:text-gray-500';

        addClassesToElement(elementId, classesToAdd);
    }
};

handleLanguageSwitch();
```

Not tested tbh, this is just chatgpt output 😂

**Comment by noworneverev** [comment]
> Cool, this works perfectly! 🎉
> 
> My only issue is that when you visit the page from a new device for the first time, no language option is selected. This might confuse users a little, esp. if we're going to host a public instance for the presentation, where they will be able to use their own devices.
> 
> Maybe something like this could fix this, by just setting the cookie based on the browser settings if it is not yet set?
> 
> ```ts
> const getUserBrowserLanguage = (): string => {
>     const userLanguage = navigator.language.split('-')[0];
>     return ['de', 'en'].includes(userLanguage) ? userLanguage : 'en';
> };
> 
> const checkLanguageCookie = (): boolean => !!document.cookie.split('; ').find((cookie) => cookie.startsWith('django_language='));
> 
> const setLanguageCookie = (language: string): void => {
>     document.cookie = `django_language=${language}; path=/;`;
> };
> 
> const addClassesToElement = (elementId: string, classes: string): void => {
>     const element = document.getElementById(elementId);
>     if (element) {
>         element.classList.add(...classes.split(' '));
>     }
> };
> 
> const handleLanguageSwitch = (): void => {
>     if (!checkLanguageCookie()) {
>         const userLanguage = getUserBrowserLanguage();
>         setLanguageCookie(userLanguage);
> 
>         const elementId = `lang-switch-${userLanguage}`;
>         const classesToAdd = 'underline pointer-events-none text-gray-300 dark:text-gray-500';
> 
>         addClassesToElement(elementId, classesToAdd);
>     }
> };
> 
> handleLanguageSwitch();
> ```
> 
> Not tested tbh, this is just chatgpt output 😂

It works. 🎉

File: ycms/static/src/js/utils/set-language.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        const classesToAdd = "underline pointer-events-none text-gray-300 dark:text-gray-500";
+
+        addClassesToElement(elementId, classesToAdd);
+    }
+};
+
+window.addEventListener("load", () => {
+    if (!checkLanguageCookie()) {
+        handleLanguageSwitch();
+    }
```
Comment (charludo) [inline ycms/static/src/js/utils/set-language.ts]: You can remove one of the `!checkLanguageCookie()` checks, schouldn't matter which

File: ycms/cms/templates/_base.html [hunk 3]
```diff
@@ -111,25 +111,44 @@
...
+                        EN
+                    </button>
+                </form>
+                <div class="border-l h-6"></div>
+                <form class="ml-2" action="{% url 'switch-language' 'de' %}" method="post">
+                    {% csrf_token %}
+                    <button type="submit"
+                            class="{% if request.COOKIES.django_language == 'de' %} underline pointer-events-none text-gray-300 dark:text-gray-500 {% endif %} text-gray-600 hover:text-blue-500 focus:outline-none dark:text-gray-50 dark:hover:text-gray-300">
+                        DE
+                    </button>
```
Comment (charludo) [inline ycms/cms/templates/_base.html]:
```suggestion
                    <button type="submit"
                            id="lang-switch-en"
                            class="{% if request.COOKIES.django_language == 'en' %} underline pointer-events-none text-gray-300 dark:text-gray-500 {% endif %} text-gray-600 hover:text-blue-500 focus:outline-none dark:text-gray-50 dark:hover:text-gray-300">
                        EN
                    </button>
                </form>
                <div class="border-l h-6"></div>
                <form class="ml-2" action="{% url 'switch-language' 'de' %}" method="post">
                    {% csrf_token %}
                    <button type="submit"
                            id="lang-switch-de"
                            class="{% if request.COOKIES.django_language == 'de' %} underline pointer-events-none text-gray-300 dark:text-gray-500 {% endif %} text-gray-600 hover:text-blue-500 focus:outline-none dark:text-gray-50 dark:hover:text-gray-300">
                        DE
                    </button>
```

:D

**Review by charludo — CHANGES_REQUESTED** [review]
Nice, thanks! Apart from the two things below + squashing the commits, this looks ready to go!

File: ycms/cms/templates/_base.html (hunk 3 above)
Comment (charludo) [reply ycms/cms/templates/_base.html]: Hmm, weird that there's no highlighting of the suggested changes. It's just the two missing IDs though

**Review by charludo — APPROVED** [review]
:tada:

**Pull request merged**